testsuite/forge/src/tx_emitter.rs

```diff
@@ -0,0 +1,717 @@
+}
+
+#[derive(Debug, Default)]
+pub struct TxStatsRate {
+    pub submitted: u64,
+    pub committed: u64,
+    pub expired: u64,
+    pub latency: u64,
+    pub p99_latency: u64,
+}
+
+#[derive(Default)]
+struct StatsAccumulator {
+    submitted: AtomicU64,
+    committed: AtomicU64,
+    expired: AtomicU64,
+    latency: AtomicU64,
+    latencies: Arc<AtomicHistogramAccumulator>,
+}
+
+struct Worker {
+    join_handle: JoinHandle<Vec<LocalAccount>>,
+}
+
+pub struct EmitJob {
+    workers: Vec<Worker>,
+    stop: Arc<AtomicBool>,
+    stats: Arc<StatsAccumulator>,
+}
+
+struct SubmissionWorker {
+    accounts: Vec<LocalAccount>,
+    client: JsonRpcClient,
+    all_addresses: Arc<Vec<AccountAddress>>,
+    stop: Arc<AtomicBool>,
+    params: EmitThreadParams,
+    stats: Arc<StatsAccumulator>,
+    tx_factory: TransactionFactory,
+}
+
+impl SubmissionWorker {
+    #[allow(clippy::collapsible_if)]
+    async fn run(mut self) -> Vec<LocalAccount> {
+        let wait = Duration::from_millis(self.params.wait_millis);
+        while !self.stop.load(Ordering::Relaxed) {
+            let requests = self.gen_requests();
+            let num_requests = requests.len();
+            let start_time = Instant::now();
+            let wait_util = start_time + wait;
+            let mut tx_offset_time = 0u64;
+            for request in requests {
+                let cur_time = Instant::now();
+                tx_offset_time += (cur_time - start_time).as_millis() as u64;
+                self.stats.submitted.fetch_add(1, Ordering::Relaxed);
+                let resp = self.client.submit(&request).await;
+                if let Err(e) = resp {
+                    warn!("[{:?}] Failed to submit request: {:?}", self.client, e);
+                }
+            }
+            if self.params.wait_committed {
+                if let Err(uncommitted) =
+                    wait_for_accounts_sequence(&self.client, &mut self.accounts).await
+                {
+                    let end_time = (Instant::now() - start_time).as_millis() as u64;
+                    let num_committed = (num_requests - uncommitted.len()) as u64;
+                    let latency = end_time - tx_offset_time / num_requests as u64;
+                    self.stats
+                        .committed
+                        .fetch_add(num_committed, Ordering::Relaxed);
+                    self.stats
+                        .expired
+                        .fetch_add(uncommitted.len() as u64, Ordering::Relaxed);
+                    self.stats.latency.fetch_add(
+                        // To avoid negative result caused by uncommitted tx occur
+                        // Simplified from:
+                        // end_time * num_committed - (tx_offset_time/num_requests) * num_committed
+                        // to
+                        // (end_time - tx_offset_time / num_requests) * num_committed
+                        latency * num_committed as u64,
```

Comment:
Should move this up into the `latency = ` statement so to keep the calculation all together, or name one `committed_latency` and `total_latency` or somethign like that.